src/apps/service/apps/de/validation.clj

```diff
@@ -45,6 +45,18 @@
                        :t.tool_id         nil
                        :t.external_app_id nil}))))
 
+(defn- remove-owned-tools
+  [username tools]
+  (let [tool-ids  (mapv :id tools)
+        owned-ids (set (keys (perms-client/load-tool-permissions username tool-ids "own")))]
+    (remove (comp owned-ids :id) tools)))
+
+(defn- get-unpublishable-tools
+  [username tools]
+  (let [public-tool-ids (perms-client/get-public-tool-ids)]
+    (->> (remove (comp public-tool-ids :id) tools)
+         (remove-owned-tools username))))
```

Comment:
We had also discussed restricting auto-publish only for tools that had a container image that was already in our private registry, so the image name had to start with `docker.cyverse.org/`.

Should we implement this restriction in these service endpoints as well?

Reply:
Ah, I'd forgotten about that requirement. Thanks for mentioning it. Yeah, that requirement should be also be in the services.

Reply:
The validation has been added. I think this is ready for another review.